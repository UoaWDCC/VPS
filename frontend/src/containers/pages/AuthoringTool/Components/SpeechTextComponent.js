import React, { useState, useContext } from "react";
import AuthoringToolContext from "context/AuthoringToolContext";
import SpeechTextboxArrow from "components/SpeechboxArrow";
import { Typography } from "@material-ui/core";
import useStyles from "./components.styles";

const borderWidth = 3.429; // px

/**
 * This React component represents a speech text scene component
 * @component
 * @example
 * <SpeechTextComponent
 *    id={index}
 *    onClick={onClick}
 *    component={component}
 * />
 */
export default function SpeechTextComponent({ id, onClick, component }) {
  const { defaultComponentStyling, speechTextComponentStyles } =
    useStyles(component);
  const [speechTextboxArrowWidth, setArrowWidth] = useState(null);
  const context = useContext(AuthoringToolContext);
  const textRef = context?.propertiesRefs?.text;
  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
    <div
      className={defaultComponentStyling}
      style={{
        display: "flex",
        flexDirection: "column",
        minWidth: speechTextboxArrowWidth + 10, // speechbox slightly wider than arrow
      }}
      onClick={onClick}
      onDoubleClick={() => textRef?.current?.focus()}
      id={id}
    >
      <SpeechTextboxArrow
        borderWidth={borderWidth}
        setArrowWidth={setArrowWidth}
      />

      <Typography
        style={{
          borderWidth,
          flex: 1,
          borderRadius: "10px",
          borderTopRightRadius: 0,
        }}
        className={speechTextComponentStyles}
      >
        {component.text}
      </Typography>
    </div>
  );
}
