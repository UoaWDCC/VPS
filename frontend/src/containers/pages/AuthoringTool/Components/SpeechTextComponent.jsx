import { Typography } from "@material-ui/core";
import SpeechTextboxArrow from "components/SpeechboxArrow";
import AuthoringToolContext from "context/AuthoringToolContext";
import { useContext, useState } from "react";
import useStyles from "./components.styles";

const borderWidth = 3.429; // px

/**
 * @param {number} borderRadius
 * @param {string} squareCorner
 * @returns {string} borderRadius CSS value
 */
function formatRadius(borderRadius, squareCorner) {
  const diffRadii = new Array(4).fill(`${borderRadius}px`);

  const position = {
    TopLeft: 0,
    TopRight: 1,
    BottomRight: 2,
    BottomLeft: 3,
  }[squareCorner];

  diffRadii[position] = "0"; // this is the square corner

  return diffRadii.join(" ");
}

/**
 * This React component represents a speech text scene component
 * @component
 * @example
 * <SpeechTextComponent
 *    id={index}
 *    onClick={onClick}
 *    component={component}
 * />
 */
export default function SpeechTextComponent({ id, onClick, component }) {
  const { defaultComponentStyling, speechTextComponentStyles } =
    useStyles(component);
  const [speechTextboxArrowWidth, setArrowWidth] = useState(null);

  const context = useContext(AuthoringToolContext);
  const textRef = context?.propertiesRefs?.text;

  // style related values
  const speechboxWidthOffset = 10;
  const arrowLocation = component.arrowLocation || "right";
  const arrowLocationStyles = {
    top: { flexDirection: "column", squareCorner: "TopRight" },
    bottom: {
      flexDirection: "column-reverse",
      squareCorner: "BottomLeft",
    },
    left: { flexDirection: "row", squareCorner: "TopLeft" },
    right: {
      flexDirection: "row-reverse",
      squareCorner: "BottomRight",
    },
  }[arrowLocation];

  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events
    <div
      className={defaultComponentStyling}
      style={{
        display: "flex",
        flexDirection: arrowLocationStyles.flexDirection,
        minWidth: speechTextboxArrowWidth + speechboxWidthOffset, // speechbox slightly wider than arrow
      }}
      onClick={onClick}
      onDoubleClick={() => textRef?.current?.focus()}
      id={id}
    >
      <SpeechTextboxArrow
        borderWidth={borderWidth}
        setArrowWidth={setArrowWidth}
        arrowLocation={arrowLocation}
      />

      <Typography
        style={{
          borderWidth,
          flex: 1,
          borderRadius: formatRadius(
            speechboxWidthOffset,
            arrowLocationStyles.squareCorner
          ),
        }}
        className={speechTextComponentStyles}
      >
        {component.text}
      </Typography>
    </div>
  );
}
