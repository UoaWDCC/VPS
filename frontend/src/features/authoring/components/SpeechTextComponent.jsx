import { Typography } from "@material-ui/core";
import SpeechTextboxArrow from "components/SpeechboxArrow/SpeechboxArrow";
import AuthoringToolContext from "context/AuthoringToolContext";
import { useContext, useState } from "react";
import useStyles from "./components.styles";
import { toRgbaString } from "../../../utils/colourUtils";

const borderWidth = 3.429; // px

/**
 * @param {number} borderRadius
 * @param {string} squareCorner
 * @returns {string} borderRadius CSS value
 */
function formatRadius(borderRadius, squareCorner) {
  const diffRadii = new Array(4).fill(`${borderRadius}px`);

  const position = {
    TopLeft: 0,
    TopRight: 1,
    BottomRight: 2,
    BottomLeft: 3,
  }[squareCorner];

  diffRadii[position] = "0"; // this is the square corner

  return diffRadii.join(" ");
}

/**
 * This React component represents a speech text scene component
 * @component
 * @example
 * <SpeechTextComponent
 *    id={index}
 *    onClick={onClick}
 *    component={component}
 * />
 */
export default function SpeechTextComponent({ id, onClick, component }) {
  const { defaultComponentStyling, speechTextComponentStyles } =
    useStyles(component);
  const [speechTextboxArrowWidth, setArrowWidth] = useState(null);

  const context = useContext(AuthoringToolContext);
  const textRef = context?.propertiesRefs?.text;

  // style related values
  const speechboxWidthOffset = 10;
  const colour = toRgbaString(component.colour);
  const arrowLocation = component.arrowLocation || "right";
  const arrowLocationStyles = {
    top: { flexDirection: "column", squareCorner: "TopRight" },
    bottom: {
      flexDirection: "column-reverse",
      squareCorner: "BottomLeft",
    },
    left: { flexDirection: "row", squareCorner: "TopLeft" },
    right: {
      flexDirection: "row-reverse",
      squareCorner: "BottomRight",
    },
  }[arrowLocation];

  return (
    <div
      className={defaultComponentStyling}
      style={{
        display: "flex",
        flexDirection: arrowLocationStyles.flexDirection,
        minWidth: speechTextboxArrowWidth + speechboxWidthOffset, // speechbox slightly wider than arrow
        zIndex: component?.zPosition || 0,
      }}
      onClick={onClick}
      onDoubleClick={() => textRef?.current?.focus()}
      id={id}
    >
      <SpeechTextboxArrow
        borderWidth={borderWidth}
        setArrowWidth={setArrowWidth}
        arrowLocation={arrowLocation}
      />

      <Typography
        style={{
          borderWidth,
          flex: 1,
          borderRadius: formatRadius(
            speechboxWidthOffset,
            arrowLocationStyles.squareCorner
          ),
          color: colour,
        }}
        className={speechTextComponentStyles}
      >
        {component.text}
      </Typography>
    </div>
  );
}
