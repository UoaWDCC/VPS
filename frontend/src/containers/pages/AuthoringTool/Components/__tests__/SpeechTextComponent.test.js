import { render } from "@testing-library/react";
import SpeechTextComponent from "../SpeechTextComponent";

test("Add speech text component snapshot test", () => {
  const { baseElement } = render(
    <SpeechTextComponent
      component={{
        type: "SPEECHTEXT",
        text: "default speech text",
        left: 50,
        top: 50,
        height: 10,
        width: 20,
      }}
    />
  );

  expect(baseElement).toMatchSnapshot();
});
