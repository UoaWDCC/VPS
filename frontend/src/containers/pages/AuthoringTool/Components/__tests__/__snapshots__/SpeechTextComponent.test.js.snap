// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add speech text component snapshot test 1`] = `
<body>
  <div>
    <div
      class="makeStyles-defaultComponentStyling-5"
      style="display: flex; flex-direction: row-reverse; min-width: 43.27833830537734px; z-index: 0;"
    >
      <div
        style="position: relative; height: 33.278338305377346px; width: 33.27833830537734px; margin-left: -3.429px; align-self: flex-end; transform: rotate(90deg);"
      >
        <div
          style="box-sizing: border-box; border-left: 33.27833830537734px solid transparent; border-bottom: 33.278338305377346px solid black; position: absolute; bottom: 0px; right: 0px;"
        />
        <div
          style="margin-right: 3.429px; box-sizing: border-box; border-left: 25px solid transparent; border-bottom: 25px solid white; position: absolute; bottom: 0px; right: 0px;"
        />
      </div>
      <p
        class="MuiTypography-root makeStyles-speechTextComponentStyles-7 MuiTypography-body1"
        style="border-width: 3.429px; flex: 1; border-radius: 10px 10px 0 10px;"
      >
        default speech text
      </p>
    </div>
  </div>
</body>
`;
